### PR TITLE
[doc] Contribute the shape and ADRs on the improvements of the Sirius Web architecture

### PR DESCRIPTION
**Pull request by sbegaudeau**
Signed-off-by: Stéphane Bégaudeau <stephane.begaudeau@obeo.fr>

# Pull request template

## General purpose
What is the main goal of this pull request?
- [ ] Bug fixes
- [ ] New features
- [ ] Documentation
- [ ] Cleanup
- [ ] Tests
- [ ] Build / releng

## Project management
- [ ] Has the pull request been added to the relevant project and milestone? (Only if you know that your work is part of a specific iteration such as the current one)
- [ ] Have the `priority:` and `pr:` labels been added to the pull request? (In case of doubt, start with the labels `priority: low` and `pr: to review later`)
- [ ] Have the relevant issues been added to the pull request?
- [ ] Have the relevant labels been added to the issues? (`area:`, `difficulty:`, `type:`)
- [ ] Have the relevant issues been added to the same project and milestone as the pull request?
- [ ] Has the `CHANGELOG.adoc` been updated to reference the relevant issues?
- [ ] Have the relevant API breaks been described in the `CHANGELOG.adoc`? (Including changes in the GraphQL API)
- [ ] In case of a change with a visual impact, are there any screenshots in the `CHANGELOG.adoc`? For example in `doc/screenshots/2022.5.0-my-new-feature.png`

## Architectural decision records (ADR)
- [ ] Does the title of the commit contributing the ADR start with `[doc]`?
- [ ] Are the ADRs mentioned in the relevant section of the  `CHANGELOG.adoc`?

## Dependencies
- [ ] Are the new / upgraded dependencies mentioned in the relevant section of the `CHANGELOG.adoc`?
- [ ] Are the new dependencies justified in the `CHANGELOG.adoc`?

## Frontend

This section is not relevant if your contribution does not come with changes to the frontend.

### General purpose
- [ ] Is the code properly tested? (Plain old JavaScript tests for business code and tests based on React Testing Library for the components)

### Typing
We need to improve the typing of our code, as such, we require every contribution to come with proper TypeScript typing for both changes contributing new files and those modifying existing files.
Please ensure that the following statements are true for each file created or modified (this may require you to improve code outside of your contribution).

- [ ] Variables have a proper type
- [ ] Functions’ arguments have a proper type
- [ ] Functions’ return type are specified
- Hooks are properly typed:
	- [ ] `useMutation<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useQuery<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useSubscription<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useMachine<CONTEXT_TYPE, EVENTS_TYPE>(…)`
	- [ ] `useState<STATE_TYPE>(…)`
- [ ] All components have a proper typing for their props
- [ ] No useless optional chaining with `?.` (if the GraphQL API specifies that a field cannot be `null`, do not treat it has potentially `null` for example)
- [ ] Nullable values have a proper type (for example `let diagram: Diagram | null = null;`)

## Backend

This section is not relevant if your contribution does not come with changes to the backend.

### General purpose
- [ ] Are all the event handlers tested?
- [ ] Are the event processor tested?
- [ ] Is the business code (services) tested?
- [ ] Are diagram layout changes tested?

### Architecture
- [ ] Are data structure classes properly separated from behavioral classes?
- [ ] Are all the relevant fields final?
- [ ] Is any data structure mutable? If so, please write a comment indicating why
- [ ] Are behavioral classes either stateless or side effect free?

### How to test this PR?
_Please describe here the various use cases to test this pull request_

- [ ] Has the Kiwi TCMS test suite been updated with tests for this contribution?

### CODE REVIEW

File: doc/adrs/125_add_the_ability_to_reuse_both_the_sirius_web_frontend_and_backend_easily.adoc [hunk 1]
```diff
@@ -0,0 +1,37 @@
+= ADR-125 - Add the ability to reuse both the Sirius Web frontend and backend easily
+
+== Context
+
+To get started with an application built on top of Sirius Web, some complex steps needs to be performed.
```
Comment (mcharfadi) [inline doc/adrs/125_add_the_ability_to_reuse_both_the_sirius_web_frontend_and_backend_easily.adoc]: need

File: doc/adrs/125_add_the_ability_to_reuse_both_the_sirius_web_frontend_and_backend_easily.adoc [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+In order to give consumers the ability to select what they want to see in their applications, we will leverage the same approach as the spring-boot project.
+In https://github.com/spring-projects/spring-boot/tree/main/spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure[its autoconfiguration support], it provides configurations and services for countless frameworks, yet those services are only scanned and added to the application context if desired by the consumers.
+
+For that, the `sirius-web-application-starter` will stop scanning everything coming from Sirius Components and will only scan core concepts instead.
+We may have to update the namespace of some Sirius Components concepts to identify them clearly as core concepts.
+Most of the Sirius Components code automatically scanned by Sirius Web should thus come `sirius-components-collaborative`, `sirius-components-emf` and their dependencies.
+We will have to review very conservatively the scope of the code which enter this module and its dependencies since they will become compulsory for every consumer of Sirius Web.
```
Comment (mcharfadi) [inline doc/adrs/125_add_the_ability_to_reuse_both_the_sirius_web_frontend_and_backend_easily.adoc]: enters

File: doc/adrs/125_add_the_ability_to_reuse_both_the_sirius_web_frontend_and_backend_easily.adoc [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+
+After that, consumers of Sirius Web will have to opt-in for every other parts of Sirius Components that they want to see included in their applications.
```
Comment (mcharfadi) [inline doc/adrs/125_add_the_ability_to_reuse_both_the_sirius_web_frontend_and_backend_easily.adoc]: part

File: doc/adrs/125_add_the_ability_to_reuse_both_the_sirius_web_frontend_and_backend_easily.adoc [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+On the frontend, things have been simplified with the introduction of `@eclipse-sirius/sirius-web-application`.
+On the backend, consumers still need to depend on every single one of our components by themselves.
+
+While consumers want to be able to reuse Sirius Web off the shelf very easily, they also want to control what can appear in the end result.
+
+== Decision
+
+We will introduce a dedicated starter for the Sirius Web backend named `sirius-web-application-starter`.
+This starter will depend on everything required to get started with both Sirius Web and Sirius Components.
+Applications depending on this starter will retrieve every feature from Sirius Components and Sirius Web at once.
```
Comment (mcharfadi) [inline doc/adrs/125_add_the_ability_to_reuse_both_the_sirius_web_frontend_and_backend_easily.adoc]: When you said every feature, do you mean only the strict minimum ?

File: doc/adrs/126_add_the_ability_to_extend_the_sirius_web_frontend.adoc [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+We also have recently introduced an API to reuse everything provided by the Sirius Web frontend very easily with the components `SiriusWebApplication`.
+While it helps get started with Sirius Web, it does not provide the same API features as hand crafting your frontend based on Sirius Components.
+While we cannot possibly provide the same level of configurability as forking the Sirius Web frontend, there are key APIs missing to help extend Sirius Web.
+
+== Decision
+
+We will introduce a uniformed API to extend the various parts of Sirius Components which will be reused to extend additional parts of Sirius Components or Sirius Web.
+We will validate this new API on both the extensibility of a Sirius Web component and an existing Sirius Components use case.
+
+To fulfill our needs, the new API should let us both replace some existing component and contribute additional components too.
```
Comment (mcharfadi) [inline doc/adrs/126_add_the_ability_to_extend_the_sirius_web_frontend.adoc]: components

**Review by mcharfadi — APPROVED** [review]

File: doc/adrs/125_add_the_ability_to_reuse_both_the_sirius_web_frontend_and_backend_easily.adoc (hunk 1 above)
Comment (sbegaudeau) [reply doc/adrs/125_add_the_ability_to_reuse_both_the_sirius_web_frontend_and_backend_easily.adoc]: Done

File: doc/adrs/125_add_the_ability_to_reuse_both_the_sirius_web_frontend_and_backend_easily.adoc (hunk 2 above)
Comment (sbegaudeau) [reply doc/adrs/125_add_the_ability_to_reuse_both_the_sirius_web_frontend_and_backend_easily.adoc]: Done

File: doc/adrs/125_add_the_ability_to_reuse_both_the_sirius_web_frontend_and_backend_easily.adoc (hunk 3 above)
Comment (sbegaudeau) [reply doc/adrs/125_add_the_ability_to_reuse_both_the_sirius_web_frontend_and_backend_easily.adoc]: Done

File: doc/adrs/125_add_the_ability_to_reuse_both_the_sirius_web_frontend_and_backend_easily.adoc (hunk 4 above)
Comment (sbegaudeau) [reply doc/adrs/125_add_the_ability_to_reuse_both_the_sirius_web_frontend_and_backend_easily.adoc]: I will make that a bit more explicit, everything will be included in the classpath but not everything will be activated by default.

**Pull request merged**